src/ErrorHandler.php

```diff
@@ -1,120 +1,93 @@
 <?php
-declare(strict_types = 1);
+declare(strict_types=1);
 
 namespace Middlewares;
 
+use Middlewares\Formatter\FormatterInterface;
+use Middlewares\Formatter\PlainFormatter;
+use Middlewares\Utils\Factory;
+use Psr\Http\Message\ResponseFactoryInterface;
 use Psr\Http\Message\ResponseInterface;
 use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Message\StreamFactoryInterface;
 use Psr\Http\Server\MiddlewareInterface;
 use Psr\Http\Server\RequestHandlerInterface;
 use Throwable;
 
 class ErrorHandler implements MiddlewareInterface
 {
-    private $handler;
+    /** @var ResponseFactoryInterface */
+    private $responseFactory;
 
-    /**
-     * @var callable|null The status code validator
-     */
-    private $statusCodeValidator;
-
-    /**
-     * @var bool Whether or not catch exceptions
-     */
-    private $catchExceptions = false;
-
-    /**
-     * @var string The attribute name
-     */
-    private $attribute = 'error';
-
-    public function __construct(RequestHandlerInterface $handler = null)
-    {
-        $this->handler = $handler;
-    }
-
-    /**
-     * Configure the catchExceptions.
-     */
-    public function catchExceptions(bool $catch = true): self
-    {
-        $this->catchExceptions = (bool) $catch;
+    /** @var StreamFactoryInterface */
+    private $streamFactory;
 
-        return $this;
-    }
-
-    /**
-     * Configure the status code validator.
-     */
-    public function statusCode(callable $statusCodeValidator): self
-    {
-        $this->statusCodeValidator = $statusCodeValidator;
+    /** @var FormatterInterface[] */
+    private $formatters = [];
 
-        return $this;
+    public function __construct(
+        ResponseFactoryInterface $responseFactory = null,
+        StreamFactoryInterface $streamFactory = null
+    ) {
+        $this->responseFactory = $responseFactory ?? Factory::getResponseFactory();
+        $this->streamFactory = $streamFactory ?? Factory::getStreamFactory();
     }
 
     /**
-     * Set the attribute name to store the error info.
+     * Add additional error formatters
      */
-    public function attribute(string $attribute): self
+    public function addFormatters(FormatterInterface ...$formatters): self
     {
-        $this->attribute = $attribute;
+        foreach ($formatters as $formatter) {
+            $this->formatters[$formatter->contentType()] = $formatter;
+        }
 
         return $this;
     }
 
-    /**
-     * Process a server request and return a response.
-     */
     public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
     {
-        ob_start();
-        $level = ob_get_level();
-
         try {
-            $response = $handler->handle($request);
-
-            if ($this->isError($response->getStatusCode())) {
```

Comment:
Why did you remove the ability to handle responses with error? (status code between 400-599)
The purpose of this middleware is to handle all http errors, not only 500.

Reply:
And what is the definition of "http error"? If a response is already defined, it should be used. What value does blowing away an existing response with an error status provide? For instance, if my API already returns 404 based on routing, why would I want to lose that and have it replaced with some generic error message?

Reply:
Not to mention things like [HTTP Problem](https://docs.zendframework.com/zend-problem-details/response/) that can be used to define a PSR-7 response with a specific error format.

Reply:
I have some webpages using this feature to display generic 404, 500, etc error pages.
A workaround could be an option to enable this feature (disabled by default), allowing to define the status code that will be considered error.

Reply:
That doesn't seem like the right approach to me.

Reply:
Again, I don't see the point of that. All of those cases (assuming they are not exceptional) should already be returning a fully-formed response with details about the error. Each of the examples listed should tell the consumer *why* the request failed. And then I go back to: why would we want to overwrite a fully-formed error response with a generic response?

Reply:
There may be cases where many different responses display the same result. An example is github: you see the same 404 error page if you enter in an inexistent user/repository (https://github.com/asofijasfo) or if the repository exists but you don't have permission to see, or are not logged. 

Sometimes I need to show the same 404 page over and over, but all there responses are created from many sites:

- The route has not found
- The route has found but the object was not found in the database
- The object has found but it's not visible or the user has not permission
- Etc.

In addition to that, there's the [HttpErrorException](https://github.com/middlewares/error-handler/blob/master/src/HttpErrorException.php) created to convert any exception into http errors. That allows to access to the exception that causes the http error in the error handler and do something with it (logging, display info to the user, etc).

Reply:
Right, I would expect HttpErrorException to be thrown in any situation where you want a generic 404 page.